File: 0701_0800/0744/0744.py
def checkstr(s):
    if len(s) % 2:
        return False
    d = {')' : '(', ']' : '[', '}' : '{'}
    t = []
    for x in s:
        if t and x in d and d[x] == t[-1]:
            t.pop()
        elif t and x in d and t[-1] in d.values() and d[x] != t[-1]:
            return False
        else:
            t.append(x)
    for i in range(len(t) // 2):
        if d[t[i]] != t[-i - 1]:
            return False
    return True

with open("input.txt", "r") as f, open("output.txt", "w") as q:
    q.write("YES" if checkstr(f.readline().strip()) else "NO")

File: 0701_0800/0744/0744.c
#include <stdio.h>
#include <string.h>

int checkstr(char *s)
{
	int j = 0;

	if (strlen(s) % 2)
		return 0;
	for (int i = 0; s[i]; ++i)
		if (j && ((s[i] == ')' && s[j - 1] == '(') || (s[i] == ']' && s[j - 1] == '[') || (s[i] == '}' && s[j - 1] == '{')))
			--j;
		else if (j && ((s[i] == ')' && (s[j - 1] == '[' || s[j - 1] == '{')) || (s[i] == ']' && (s[j - 1] == '(' || s[j - 1] == '{')) || (s[i] == '}' && (s[j - 1] == '[' || s[j - 1] == '('))))
			return 0;
		else
			s[j++] = s[i];
	for (int i = 0; i < j / 2; ++i)
		if (!((s[i] == ')' && s[j - 1 - i] == '(') || (s[i] == ']' && s[j - 1 - i] == '[') || (s[i] == '}' && s[j - 1 - i] == '{')))
			return 0;
	return 1;
}

int main()
{
	FILE *f = fopen("input.txt", "r"), *q = fopen("output.txt", "w");
	char s[10001];

	fscanf(f, "%s", s);
	fprintf(q, checkstr(s) ? "YES" : "NO");
	return 0;
}
